### ISSUE TITLE
Wrong error message with closures on nightly

### ISSUE THREAD

**Issue by mcarton** [rust-lang/rust#42143]
The following example

```rust
fn main() {
    let a = vec![(1, 2)];
    
    let b: Vec<_> = a.iter().map(|x: (u32, u32)| 45).collect(); 
}
```

gives the following error on nightly:

```
error[E0593]: closure takes 2 arguments but 1 argument is required
 --> src/main.rs:4:30
  |
4 |     let b: Vec<_> = a.iter().map(|x: (u32, u32)| 45).collect(); 
  |                              ^^^ ------------------ takes 2 arguments
  |                              |
  |                              expected closure that takes 1 argument
```

which is wrong (`map` is expecting a function with one argument, which is what this closure is) and the following correct message on stable/beta:

```
error[E0281]: type mismatch: the type `[closure@src/main.rs:4:34: 4:52]` implements the trait `std::ops::FnMut<((u32, u32),)>`, but the trait `std::ops::FnMut<(&({integer}, {integer}),)>` is required (expected reference, found tuple)
 --> src/main.rs:4:30
  |
4 |     let b: Vec<_> = a.iter().map(|x: (u32, u32)| 45).collect(); 
  |                              ^^^
```

**Comment by rbalicki2** [comment]
How does one fix this? I have run into this problem as well, and have tried all 8 variants of `|&x: &(&u32, &u32)|`, toggling each of the 3 placements of ampersands.

It looked like, at some point, that `.map(|x: &(i32, i32)| 42)` fixed it, but I could not get that to work for `group_by` of `itertools`. Perhaps these are separate issues.

**Comment by estebank** [comment]
Introduced in #41488.

**Comment by Emilgardis** [comment]
I would like to tackle this if no one already has begun doing it.

**Comment by estebank** [comment]
@Emilgardis go ahead! You should add a new test for this case, and you'll need to add an extra check somewhere in [`report_selection_error`](https://github.com/rust-lang/rust/blob/master/src/librustc/traits/error_reporting.rs#L679-L728). Don't hesitate to ask for help either here, in IRC or https://internals.rust-lang.org.

**Comment by Emilgardis** [comment]
@estebank Do I add type_mismatch .rs test or add to an existing test file?

**Comment by Emilgardis** [comment]
pr on #42270
